### PR TITLE
Add json validator

### PR DESCRIPTION
**Pull request by romkatsu**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | 

- add json validator
- add trait
- edit composer.json

### CODE REVIEW

File: src/HasValidationMessage.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?php
```
Comment (xepozz) [inline src/HasValidationMessage.php:1]: I don't sure that this trait is really useful. Maybe we can place the `message` method into parent `Rule` instead?

File: src/Rule/Json.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    private function isValidJson($value): bool
+    {
+        if (!is_string($value)) {
+            return false;
+        }
+
+        try {
+            json_decode($value, false, 512, JSON_THROW_ON_ERROR);
+            return true;
+        } catch (Exception $e) {
```
Comment (xepozz) [inline src/Rule/Json.php:42]:
```suggestion
        } catch (\JsonException $e) {
```

File: src/HasValidationMessage.php (hunk 1 above)
Comment (romkatsu) [reply src/HasValidationMessage.php:1]: Yes, we can. But not all validators will use it. So I put it in the trait.

File: src/Rule/Json.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    private function isValidJson($value): bool
+    {
+        if (!is_string($value)) {
+            return false;
+        }
+
+        try {
+            json_decode($value, false, 512, JSON_THROW_ON_ERROR);
+            return true;
```
[the remark this answers is not in the document]
Comment (romkatsu) [reply src/Rule/Json.php]: done

File: tests/Rule/JsonTest.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        $this->assertFalse($val->validate(10)->isValid());
+        $this->assertFalse($val->validate(null)->isValid());
+    }
+
+    public function testValidValueValidate(): void
+    {
+        $this->assertTrue((new Json())->validate('{"name": "tester"}')->isValid());
+    }
+
+    public function testValidationMessage(): void
```
[the remark this answers is not in the document]
Comment (romkatsu) [reply tests/Rule/JsonTest.php:35]: done

File: src/HasValidationMessage.php (hunk 1 above)
Comment (xepozz) [reply src/HasValidationMessage.php:1]: Which ones?

File: src/Rule/Json.php (hunk 2 above)
Comment (romkatsu) [reply src/Rule/Json.php]: done

File: src/HasValidationMessage.php (hunk 1 above)
Comment (romkatsu) [reply src/HasValidationMessage.php:1]: Callback, CompareTo, Ip, Number

Comment (xepozz) [reply src/HasValidationMessage.php:1]: Ok thanks

**Comment by romkatsu** [comment]
[Json validator](https://trello.com/c/hJAlXABg/58-json-validator)

**Comment by samdark** [comment]
👍

**Comment by armpogart** [comment]
Not sure if it really matters but I have a benchmark (if I implemented it correctly) that shows that using regex for validating json is faster, and the bigger the json, the more is the difference. For real world case the difference would not be visible at all, but another disadvantage of using `json_decode` for validation purposes here is that it will also consume memory by creating all that objects for the decoded data.

See [benchmark](http://sandbox.onlinephpfunctions.com/code/afc77f800197659391c3d1ea2a02b5e1884b9a4f)

**Comment by armpogart** [comment]
Also `json_decode` successfully decodes any number (which you have check for), simple strings and boolean true (which also don't pass `is_string` check), which is not valid json in most languages. But simple strings still pass you validation, but they are not valid for javascript `JSON.parse`.

That's due to the fact that PHP implements a superset of JSON specified in [RFC 7159](https://www.ietf.org/rfc/rfc7159.txt) vs older standard that is implemented by javascript and majority: [RFC 4267](https://www.ietf.org/rfc/rfc4627.txt).

Also there is a newer revision of json standard: [RFC 8259](https://tools.ietf.org/html/rfc8259), which I doubt any language implemented at all.

**Comment by samdark** [comment]
@armpogart are you sure it is complete regex that validates JSON?

**Comment by armpogart** [comment]
@samdark I'll open another discussion (issue) with details posted here and references to all the data I got. There are also json validation test cases on [json.org](https://www.json.org/json-en.html) which can be used for json validator testing.

Anyways I'll post all the details in an issue reference it here. I think there are some points that need further discussion.

**Comment by samdark** [comment]
Alright. Overall benchmarks look good enough to try it.

**Pull request merged**